### PR TITLE
Expand auto local access optimization to support basic offsets 

### PR DESCRIPTION
**Pull request by e-kayrakli**
This PR expands automatic localAccess (ALA) optimization to cover accesses with offsets. Currently, this is only applicable to arrays with StencilDist.

Indexing operations such as the following are supported:

```chpl
const constOff = 1;
param paramOff = 1;
forall i in ... {
  Arr[i+1]
  Arr[i-1]
  Arr[i+paramOff]
  Arr[i-constOff]
  Arr[1-i]
  Arr[-1+i]
  // not an exhaustive list of all combos
  // multidimensional indexing is also supported
}
```

### Implementation Overview

- `proc dsiSupportsOffsetAutoLocalAccess param` on domains control whether the type supports this optimization. It is off by default on all domains, it is only on by default in StencilDist.
  - Calls to this function is added to static checks that control ALA.
  - Note that we keep track of static checks with and without offsets separately. Otherwise ALA might get thrown off for things like `MyBlockArr[i]` and `MyBlockArr[i+1]`, where the failure of the latter could thwart optimizing the former. 
- `dsiAutoLocalAccessOffsetCheck(offsets)` implements the dynamic check to see whether the tuple `offsets` can be used to locally access the array. Practically, this checks whether absolute value of all elements of `offsets` is less than the corresponding element in the `fluff` tuple of Stencil domains.
  - We call this function as part of dynamic checks.
  - Note that offset-accesses are by definition dynamic candidates. `fluff` is not a param entity, so we can't optimize offset-accesses statically.
- `--no-offset-auto-local-access` can be used to disable this optimization. 

### Some more details

- This PR creates the type `ALACandidate`. This type is used to keep track of ALA candidates, whereas before we were just keeping track of things like calls as we didn't have to carry a lot of information from call to call. This type allows us to keep track of offsets efficiently without having to recompute them etc. I believe there's more code that can be folded into this type's members. But I didn't want to spiral out too much.
- I came across a bug in auto aggregation (AA) that's probably not very possible to see before this PR. It is fixed now.
- A lot of tests for ALA and AA has their good files changed. None shows execution-time differences. With this PR `Arr[i+x]` expressions are also subject to ALA optimization and even though in all cases they are not optimizable (`Arr` needs to be Stencil, after all), optimized branches for such loops exist and live from normalize to resolve. Note that AA is built on top of ALA, so they are tightly related. Thus, they impact the output from `--report-auto-local-access` and `--report-auto-aggregation`
- New tests for this are in `test/optimization/autoLocalAccess/stencil`.

Resolves https://github.com/chapel-lang/chapel/issues/25689

Test:
- [x] linux64
- [x] gasnet

### CODE REVIEW

File: compiler/include/ForallStmt.h [hunk 1]
```diff
@@ -31,6 +31,67 @@ enum ForallAutoLocalAccessCloneType {
...
+    ALACandidate() = delete;
+    ALACandidate(CallExpr *call, ForallStmt *forall, bool checkArgs=false);
+    inline CallExpr* getCall() const { return call_; }
+
+    inline int getIterandIdx() const { return iterandIdx_; }
+    inline void setIterandIdx(int iterandIdx) { iterandIdx_ = iterandIdx; }
+
+    inline CallRejectReason getReason() const { return reason_; }
+    inline void setReasonIfNeeded(CallRejectReason reason) {
+      if (!hasReason()) reason_ = reason;
```
Comment (benharsh) [inline compiler/include/ForallStmt.h:58]: Should this check for CRR_ACCEPT? If it was previously accepted and a failure reason is passed in, it won't be set.

**Review by benharsh — APPROVED** [review]

File: compiler/include/ForallStmt.h (hunk 1 above)
Comment (e-kayrakli) [reply compiler/include/ForallStmt.h:58]: Hmm, while I can't think of a way that's happening in current implementation, at a high-level that feels safe, and easy to add. I'll do it.

Comment (e-kayrakli) [reply compiler/include/ForallStmt.h]: https://github.com/chapel-lang/chapel/pull/25712/commits/a2485dd6d2236f3c209c6cf40a5e8e249cd40a32

**Comment by e-kayrakli** [comment]
I addressed @benharsh's feedback and adjusted the remaining good files. Running the last round of paratest before merging.

**Comment by bradcray** [comment]
@e-kayrakli : Do we have performance tests that will demonstrate an improvement due to this?  I was wondering whether PRK Stencil would (e.g., maybe the elegant version), but am not seeing anything that looks changed:  https://chapel-lang.org/perf/16-node-apollo-hdr/?startdate=2024/06/16&enddate=2024/08/14&graphs=prkstencilvariationsperf,prkoptimizedstenciltimesec,prkelegantstenciltimesec,2dheatsolvers5pointstencil

**Comment by e-kayrakli** [comment]
You're mostly right. First, if you zoom in on the two fastest curves of "2D Heat Solvers", you'll see that the localAccess vs non-localAccess versions perform nearly identically in the last run. I believe that's due to this PR.

PRK-Stencil doesn't benefit from this PR, unfortunately. I don't think it is an easy lift either. A quick snippet:

```chpl
forall (i,j) in innerDom with (const in weight, ref output) {
  ..
  for param jj in -R..-1 do
    tmpout += weight[R][R+jj] * input[i, j+jj];
```

Here we want to optimize `input`. But recall that ALA analysis fires right before normalization. At that point, `param` loops are not lowered yet. Therefore, according to ALA, the `+jj` offset is not constant enough. We could consider deferring such cases to post resolution, but how that can be implemented is not immediately obvious to me.

**Comment by bradcray** [comment]
> you'll see that the localAccess vs non-localAccess versions perform nearly identically in the last run. I believe that's due to this PR.

Oh, interesting.  Though they've re-diverged today (and seemingly by a wider margin).  :(  Looks like we need more time/data still.

Thanks for the PRK Stencil explanation!  I had a vague memory that something about its stencils were different, but couldn't remember what (and didn't take the time to look into it).  And then wouldn't have remembered the timing of the passes even if I had.

**Comment by e-kayrakli** [comment]
> Oh, interesting. Though they've re-diverged today (and seemingly by a wider margin). :( Looks like we need more time/data still.

Hmm, yes, I see today's data. The test is noisy and even today's relatively slower run with the elegant version is faster than any other data point for that same version. So, I am hopeful that this signal and not noise.

Note that we may never see the elegant and localAccess curves juxtapose for that benchmark. For offsets, we will need dynamic checks as long as we don't have param tuples or some other solutions to support static fluff size in StencilDist. For a quick and iterative benchmark like this, that cost can add up.

**Comment by bradcray** [comment]
> and even today's relatively slower run with the elegant version is faster than any other data point for that same version.

Oh good point, I was looking at the segment=large lines rather than the segment=fast lines.

**Comment by e-kayrakli** [comment]
Thinking about PRK-Stencil this morning, I think there's a quite acceptable, almost natural solution that's incremental and doesn't require a big overhaul of how ALA is implemented:

```chpl
forall (i,j) in Arr.domain {
  for offset in x..y {
    Arr[i+offset,j-offset];
  }
}
```

In this case, all we need to do is to extract `x` and `y` from the inner loop, make sure that they are constant within the loop's body and use them as offsets. Because fundamentally, all we need is to find the largest offset (we do that by `max`ing all the offsets we have) in each dimension and loop bounds can be representative of that.

The most challenging part is proving constness within the loop body. But we don't need that tight of a check: if they are literals, params or consts, we can work with that. For PRK-Stencil, this inner loop is a `param` loop. So they have to be literals or params.

This feels very natural because ALA revolves around similar static analysis where domains of arrays/loops are tracked down. Here, we'd be tracking down the bounds of `offset`.

Implementation-wise, this is probably 2-3 points for me. I am not sure whether I can/should afford that for 2.2, though. As much as I'd enjoy writing this code, I'd be also more than happy to introduce others to the code and help them implement this :)

**Pull request merged**